api/provider/aws/deploy_create.go

```diff
@@ -0,0 +1,131 @@
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+func (d *DeployProvider) Create(req models.CreateDeployRequest) (*models.Deploy, error) {
+	if req.DeployName == "" {
+		return nil, fmt.Errorf("DeployName is required")
+	}
+
+	deploy, err := d.createDeploy(req.DeployName, req.Dockerrun)
+	if err != nil {
+		return deploy, err
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "name", Value: req.DeployName}); err != nil {
+		return deploy, err
+	}
+
+	if err := d.TagStore.Insert(models.Tag{EntityID: deploy.DeployID, EntityType: "deploy", Key: "version", Value: deploy.Version}); err != nil {
+		return deploy, err
+	}
+
+	return deploy, nil
+}
+
+func (d *DeployProvider) createDeploy(deployName string, body []byte) (*models.Deploy, error) {
+	// since we use '.' as our ID-Version delimiter, we don't allow it in deploy names
+	if strings.Contains(deployName, ".") {
+		return nil, fmt.Errorf("Deploy names cannot contain '.'")
+	}
+
+	dockerrun, err := d.createRenderedDockerrun(body)
+	if err != nil {
+		return nil, err
+	}
+
+	familyName := L0DeployID(d.Config.Instance(), deployName)
+	if dockerrun.Family != "" && dockerrun.Family != familyName {
+		return nil, fmt.Errorf("Custom family names are currently unsupported in Layer0")
+	}
+
+	taskDef := &ecs.RegisterTaskDefinitionInput{}
+	taskDef.SetFamily(familyName)
+	taskDef.SetTaskRoleArn(dockerrun.TaskRoleARN)
+	taskDef.SetNetworkMode(dockerrun.NetworkMode)
+	taskDef.SetContainerDefinitions(dockerrun.ContainerDefinitions)
+	taskDef.SetVolumes(dockerrun.Volumes)
+	// TODO: Understand why our model had to be updated to use ecs.TaskDefinitionPlacementConstraint
+	// rather than ecs.PlacementConstraint
+	taskDef.SetPlacementConstraints(dockerrun.PlacementConstraints)
+
+	return d.populateModel(taskDef)
+}
+
+func (d *DeployProvider) populateModel(taskDef *ecs.RegisterTaskDefinitionInput) (*models.Deploy, error) {
```

Comment:
This function name and argument type isn't correct. I think it would make more sense to simply marshal the deploy's `*ecs.TaskDefintition` and place that into the body. This could probably be done in the "master" function:

```
taskDefinition, err := d.createTaskDefinition(...)
...

bytes, err := json.Marshal(taskDefinition)
....

model.Dockerrun = bytes
```